### PR TITLE
feat: Add refurb linter

### PR DESCRIPTION
**Pull request by q0w**
https://github.com/dosisod/refurb

### CODE REVIEW

File: lintrunner_adapters/adapters/refurb_linter.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    >>> t(r"main.py:4:5 [FURB101]: Use `y = Path(x).read_text()` instead of `with open(x, ...) as f: y = f.read()`")
+    ... # doctest: +NORMALIZE_WHITESPACE
+    {'file': 'main.py', 'line': '4', 'column': '5', 'code': 'FURB101',
+     'message': 'Use `y = Path(x).read_text()` instead of `with open(x, ...) as f: y = f.read()`'}
+    """
+    pass
+
+
+def format_lint_message(message: str, code: str, show_disable: bool) -> str:
+    formatted = f"{message}\n"
```
Comment (justinchuby) [inline lintrunner_adapters/adapters/refurb_linter.py:48]: is it possible to move the newline to before `To disable, use...`?

File: lintrunner_adapters/adapters/refurb_linter.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+
+def main() -> None:
+    parser = argparse.ArgumentParser(
+        description=f"refurb wrapper linter. Linter code: {LINTER_CODE}",
+        fromfile_prefix_chars="@",
+    )
+    parser.add_argument(
+        "--config-file",
+        required=True,
```
Comment (justinchuby) [inline lintrunner_adapters/adapters/refurb_linter.py:119]: Possible to default to `pyproject.toml`?

File: lintrunner_adapters/adapters/refurb_linter.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+                match["message"],
+                match["code"],
+                show_disable,
+            ),
+            line=int(match["line"]),
+            char=int(match["column"])
+            if match["column"] is not None and not match["column"].startswith("-")
+            else None,
+            code=LINTER_CODE,
+            severity=severities.get(match["code"], LintSeverity.ERROR),
```
Comment (justinchuby) [inline lintrunner_adapters/adapters/refurb_linter.py:104]: Possible to change default severity to advice? It seems more fitting for this tool

File: lintrunner_adapters/adapters/refurb_linter.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    )
+    parser.add_argument(
+        "--config-file",
+        required=True,
+        help="path to pyproject.toml config file",
+    )
+    parser.add_argument(
+        "--severity",
+        action="append",
+        help="map code to severity (e.g. `FURB109:advice`)",
```
Comment (justinchuby) [inline lintrunner_adapters/adapters/refurb_linter.py:125]: What should users input if they wanted to map more than one codes? Would be very nice to show in the example

File: requirements-test.txt [hunk 5]
```diff
@@ -1,4 +1,5 @@
 # Test dependencies to avoid doctest import errors
 add-trailing-comma==2.4.0
 pyupgrade==3.3.1
+refurb==1.10.0;python_version>="3.10"
```
Comment (justinchuby) [inline requirements-test.txt:4]: I wonder how the tests are successful with python 3.9. Do you know?

**Review by justinchuby — COMMENTED** [review]
Thank you, and this looks like a cool linter we should enable in this repo. Could you help adding the lintrunner config to this repo too?

File: lintrunner_adapters/adapters/refurb_linter.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    )
+    parser.add_argument(
+        "--show-disable",
+        action="store_true",
+        help="show how to disable a lint message",
+    )
+    add_default_options(parser)
```
Comment (justinchuby) [inline lintrunner_adapters/adapters/refurb_linter.py:132]: Should we add the `--python-version` option too?

File: lintrunner_adapters/adapters/refurb_linter.py (hunk 4 above)
Comment (q0w) [reply lintrunner_adapters/adapters/refurb_linter.py:125]: Use multiple --severity arg

File: requirements-test.txt (hunk 5 above)
Comment (q0w) [reply requirements-test.txt:4]: Sorry but i dont see any tests except doctests. wdym?

File: lintrunner_adapters/adapters/refurb_linter.py (hunk 6 above)
Comment (q0w) [reply lintrunner_adapters/adapters/refurb_linter.py:132]: Refurb supports all cli args in pyproject.toml, So I dont think there is need to also parse refurb cli args

**Comment by q0w** [comment]
> Thank you, and this looks like a cool linter we should enable in this repo. Could you help adding the lintrunner config to this repo too?

Refurb is pretty slow compared with compiled mypy. https://github.com/dosisod/refurb/issues/57. What do you think?

File: lintrunner_adapters/adapters/refurb_linter.py (hunk 1 above)
Comment (q0w) [reply lintrunner_adapters/adapters/refurb_linter.py:48]: It already does that
image

File: requirements-test.txt (hunk 5 above)
Comment (justinchuby) [reply requirements-test.txt:4]: Ah i see - refurb was not directly imported in the module unlike pu upgrade for example. So when pytest reads the module for doctest it doesn’t encounter an import error

File: lintrunner_adapters/adapters/refurb_linter.py (hunk 6 above)
Comment (justinchuby) [reply lintrunner_adapters/adapters/refurb_linter.py:133]: SG! Could you add this info to the main help message so ppl are aware?

File: lintrunner_adapters/adapters/refurb_linter.py (hunk 1 above)
Comment (justinchuby) [reply lintrunner_adapters/adapters/refurb_linter.py:48]: So I wonder if there will be an extra newline when show_disable is False. But that’s ok I suppose? I don’t know if lintrunner will trim the final newline

**Comment by justinchuby** [comment]
> > Thank you, and this looks like a cool linter we should enable in this repo. Could you help adding the lintrunner config to this repo too?
> 
> Refurb is pretty slow compared with compiled mypy. [dosisod/refurb#57](https://github.com/dosisod/refurb/issues/57). What do you think?

In that case let’s merge this first and I’ll experiment with it

File: lintrunner_adapters/adapters/refurb_linter.py (hunk 1 above)
Comment (q0w) [reply lintrunner_adapters/adapters/refurb_linter.py]: Added an extra newline to the disable message

**Comment by q0w** [comment]
How add refurb only for python3.10+ ?

**Comment by q0w** [comment]
Why lintrunner does not fail if there is no tool available in PATH? Like here pip install refurb only for py>=3.10

**Comment by justinchuby** [comment]
I will need to look more deeply

**Comment by justinchuby** [comment]
Looks like its because run_command needs check=True

**Pull request merged**